### DISCUSSION TITLE
Cannot understand how the puma plugin is getting priority

### DISCUSSION THREAD

**Discussion by vamsipavanmahesh** [puma/puma#3115] [Q&A]
So we have a simple end point `/hello` as a puma plugin. I am just trying to figure out how it's working.

This plugin is configured to have a single thread. It does not do anything other than responding with a simple hello message.

As per my understanding, the worker thread picks up one of the Threads from the pool whenever there is an I/O. But when it's CPU intensive, the threads should not be switching.

For my sample app, I have min and max puma threads as 2 and this plugin has one more thread which is on a different port.

**Problem**

Regardless of whether cpu intensive work is going or i/o intensive work is going, and the given threads are already busy, the /hello end point is always executing.

Which is strange to me, the behavior is like as if this is working on a different process on it's own. Even though when I print the PID, it's showing the same process id as puma's main process.

I want to understand how this plugin is getting priority to execute, even though a CPU intensive task is going on

Sample controller code:

```ruby
class HelloController < ApplicationController
  def cpu
    sum = 0
    for i in 1..1000
      for j in 1..1000
        for k in 1..1000
          sum += i + j + j
        end
      end
    end

    render json: {sum: sum}
  end
end
```

In the below screenshot we can observe that `/health` executed, even though 2 threads are busy with cpu intensive work.
<img width="1728" alt="Screenshot 2023-03-31 at 8 30 08 PM" src="https://user-images.githubusercontent.com/1875074/229157176-9fa33044-18aa-453b-9cb9-45f8955c8f2c.png">

**Comment by dentarg** [comment]
I think a complete repro would be good for anyone to help with an answer

**Reply by vamsipavanmahesh** [reply]
Okay, I will try to come up with a sample repo, need to figure out the plugin thing though as I am not the original author of the plugin.

But in theory, what I said makes sense? or are there any gaps in it

**Reply by vamsipavanmahesh** [reply]
are there resources i can follow to create a plugin which responds with simple json? this private plugin i am using is proprietary gem and has a lot more going on

**Reply by vamsipavanmahesh** [reply]
I looked a little further into the plugin, it looks like it's starting a new server instance ` Puma::Server.new` with a tiny rack app which is responsible for /hello

So main puma process is running in Single mode, and this plugin also seems to be adding one more Server instance, even though the main ruby process is same for both the regular app and plugin, which i verified using pid.

Since this application is using ruby 3.1, i thought may be different ractors are responsible for each of these rack apps, but the Ractor object id also turned out to be same for application and the plugin.

@nateberkopec any thoughts on this? 🙏

**Comment by nateberkopec** [comment]
>But when it's CPU intensive, the threads should not be switching.

MRI Ruby interrupts threads every 100 milliseconds. There is no way to disable this. There are some APIs for thread priority but I don't know if they really work or will do what you want.
